src/main/java/net/minecraftforge/event/entity/living/MobSpawnEvent.java

```diff
@@ -0,0 +1,238 @@
+         * This is how entities like horses ensure that the whole group spawns as a single variant. How this is used varies on a per-entity basis.
+         * @return The spawn group data.
+         */
+        @Nullable
+        public SpawnGroupData getSpawnData()
+        {
+            return this.spawnData;
+        }
+
+        /**
+         * Sets the spawn data for this entity. If this event is cancelled, this value is not used, since {@link Mob#finalizeSpawn} will not be called.
+         * @param data The new spawn data
+         * @see FinalizeSpawn#getSpawnData
+         */
+        public void setSpawnData(@Nullable SpawnGroupData data)
+        {
+            this.spawnData = data;
+        }
+
+        /**
+         * This is the NBT data the entity was loaded from, if applicable. It is unknown if the entity has already been loaded from this data, or if it will be loaded later.
+         * You should not modify this data.  If you need to change the data, you can create a copy, modify it, and set it via {@link FinalizeSpawn#setSpawnTag}
+         * @return The spawn data this entity was or will be loaded from, if any.
+         */
+        @Nullable
+        public CompoundTag getSpawnTag()
+        {
+            return this.spawnTag;
+        }
+
+        /**
+         * Sets the spawn data for this event, which will be propagated to {@link Mob#finalizeSpawn} unless cancelled.
+         * The only vanilla mob known to use this tag for anything in finalize is tropical fish for setting the variant when spawned via bucket.
+         * @param tag The new spawn tag
+         */
+        public void setSpawnTag(@Nullable CompoundTag tag)
+        {
+            this.spawnTag = tag;
+        }
+
+        /**
+         * Retrieves the underlying {@link BaseSpawner} instance if this mob was created by a Mob Spawner of some form.
+         * This is always null unless {@link #getSpawnType()} is {@link MobSpawnType#SPAWNER}, and may still be null even then.
+         * @return The BaseSpawner responsible for triggering the spawn, or null if none is available.
+         */
+        @Nullable
+        public BaseSpawner getSpawner()
+        {
+            return spawner;
+        }
+
+        /**
+         * This method can be used to fully cancel the spawn of this mob.<p>
+         * This method must be used if you want to block the spawn, as canceling the event only blocks the call to {@link Mob#finalizeSpawn}.<p>
+         * Note that if this is cancelled, but the event is not, than {@link Mob#finalizeSpawn} will still be called, but the entity will not be spawned.
+         * Usually that has no side effects, but callers should be aware.
+         * @param cancel If the spawn should be cancelled (or not).
+         */
+        public void setSpawnCancelled(boolean cancel) {
+        	this.getEntity().setSpawnCancelled(cancel);
```

Comment:
Formatting: 4 spaces, not tabs.
```suggestion
            this.getEntity().setSpawnCancelled(cancel);
```